### PR TITLE
Mesh splitting and selection

### PR DESCRIPTION
**Pull request by PProfizi**
Introduces `Meshes`, a container for multiple `Mesh` instances (wraps a `MeshesContainer`).

The goal is to allow for lazy `Mesh` queries, where a first splitting step is used with a further selection to work only on the `MeshedRegion` of interest.

Adds `Mesh.split_by_properties()`, `Mesh.select()`, `Mesh.plot()`
Adds `Meshes`, `Meshes[{"mat": 1, "elshape":1}]` or `Meshes[0]`, `Meshes.plot()`

To do:
- [x] Treat split by with specific property values
- [x] Improve coverage
- [x] Improve documentation
- [x] Add example

### CODE REVIEW

File: src/ansys/dpf/post/connectivity.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    @property
+    def by_id(self) -> ConnectivityList:
+        if self._mode == Mode.IDS_FROM_IDX:
+            return ConnectivityList(self._field, self._scoping, Mode.IDS_FROM_ID)
+        elif self._mode == Mode.IDX_FROM_IDX:
+            return ConnectivityList(self._field, self._scoping, Mode.IDX_FROM_ID)
+        return self
+
+    def _to_ids(self, indices) -> List[int]:
+        to_id = self._scoping.id
```
Comment (cbellot000) [inline src/ansys/dpf/post/connectivity.py:53]: see LocalScoping

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+# Split the global mesh and select meshes for specific property values
+print(meshes)
+meshes = simulation.split_mesh_by_properties(
+    properties={
+        elemental_properties.material: 1,
+        elemental_properties.element_shape: [0, 1],
+    }
+)
+
+# Mesh<index=0, num_nodes=100, num_elem=1000>
```
Comment (cbellot000) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:56]: Since this is an example shouldn't we remove unnecessary comments?

Comment (cbellot000) [reply examples/01-Detailed-Examples/05-mesh-exploration.py:56]: Could you also make sure that we display properly the results of each API call in the generated doc?

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
+"""
+.. _ref_mesh_exploration_example:
+
+Explore the mesh
+================
+In this script a static simulation is used as an example to show how to
+query mesh information such as connectivity, element IDs, element types and so on.
+"""
+from __future__ import annotations
```
Comment (PProfizi) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:9]: Hi @ansys-akarcher, qucik question, why is that needed here?

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+from ansys.dpf import post
+from ansys.dpf.post import examples
+from ansys.dpf.post.common import elemental_properties
+
+###############################################################################
```
Comment (PProfizi) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:15]: @ansys-akarcher I think this header is supposed to be above the imports

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+# Get the mesh
+# ------------
+
+# Retrieve the actual mesh
+mesh = simulation.mesh
+
+###############################################################################
+# Query basic information about the mesh (available in PyDPF-Core)
+# --------------------------------------
```
Comment (PProfizi) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:47]:
```suggestion
# Query basic information about the mesh
# --------------------------------------
```

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+# Get the nodes of an element
+mesh.elements[1].n_nodes
+
+# Get the type of the element
+mesh.elements[1].type_info
+mesh.elements[1].type_id
+
+# Get the shape of the element
+mesh.elements[1].shape
```
Comment (PProfizi) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:132]: @ansys-akarcher It may be a better idea to illustrate here that mesh.elements[] is by index, thus zero-based.
Illustrating with "1" means there can be a doubt IMO.
```suggestion
mesh.elements[0].nodes

# Get the node IDs of an element
mesh.elements[0].node_ids

# Get the nodes of an element
mesh.elements[0].n_nodes

# Get the type of the element
mesh.elements[0].type_info
mesh.elements[0].type_id

# Get the shape of the element
mesh.elements[0].shape
```

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+# elements indices to elem_id
+print(conn3[0])
+
+# elements IDs from node index
+print(conn4[0])
+
+###############################################################################
+# Splitting into meshes
+# --------
```
Comment (PProfizi) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:182]: Underlines must be of the same length
```suggestion
# Splitting into meshes
# ---------------------
```

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+# Split the global mesh and select meshes for specific property values
+print(meshes)
+meshes = simulation.split_mesh_by_properties(
+    properties={
+        elemental_properties.material: 1,
+        elemental_properties.element_shape: [0, 1],
+    }
+)
+
+# Mesh<index=0, num_nodes=100, num_elem=1000>
```
Comment (PProfizi) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:202]:
```suggestion
```

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+meshes.plot()
+
+# Select a specific Mesh in the Meshes, by index
+meshes[1].plot()
```
Comment (PProfizi) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:206]: Again, when working by index, it might be better to illustrate it using "0" so we know it is unequivocally zero-based
```suggestion
meshes[0].plot()
```

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+# --------------------------------------
+
+# Node IDs
+n_ids = mesh.node_ids
+
+# Element IDs
+e_ids = mesh.element_ids
+
+# Available named selection names
+named_selections = mesh.available_named_selections
```
Comment (cbellot000) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:56]: Can we remove available_named_selections if we have named_selections with all the info please?

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+    print(v)
+for k, v in mesh.named_selections.items():
+    print(f"{k} = {v}")
+
+###############################################################################
+# Get elements
+# ------------
+
+# Get an element by ID
+el_by_id = mesh.elements.by_id[1]
```
Comment (cbellot000) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:96]: for ease of use we can have get_element_by_id too

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+# ----------------------------------------------
+
+# Get the nodes of an element
+mesh.elements[0].nodes
+
+# Get the node IDs of an element
+mesh.elements[0].node_ids
+
+# Get the nodes of an element
+mesh.elements[0].n_nodes
```
Comment (cbellot000) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:125]: we would rather see num_nodesand num_elements for consistency

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+
+# Get the type of the element
+mesh.elements[0].type_info
+mesh.elements[0].type_id
```
Comment (cbellot000) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:129]: type is clearer than type_id

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+
+###############################################################################
+# Query information about one particular node
+# -------------------------------------------
+
+# Coordinates
+mesh.nodes[1].coordinates
+
+# Get Nodal connectivity
+conn3 = mesh.conn_node_to_elem
```
Comment (cbellot000) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:169]: after discussion, we can rename to element_to_node_connectivity and element_to_nodes_ids_connectivity

File: examples/01-Detailed-Examples/05-mesh-exploration.py [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+query mesh information such as connectivity, element IDs, element types and so on.
+"""
+
+###############################################################################
+# Perform required imports
+# ------------------------
+# Perform required imports. # This example uses a supplied file that you can
+# get by importing the DPF ``examples`` package.
+
+from ansys.dpf import post
```
Comment (cbellot000) [inline examples/01-Detailed-Examples/05-mesh-exploration.py:16]: as we discussed @ansys-akarcher, in my opinion this is not really an example, it's more an api display, if you could move these to the mesh docstrings, it would be great

File: src/ansys/dpf/post/connectivity.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+
+    def __init__(
+        self,
+        field: PropertyField,
+        mode: ReturnMode,
+        scoping: Scoping,
+    ):
+        """Constructs a _ConnectivityList by wrapping the given PropertyField.
+
+        Arguments:
```
Comment (cbellot000) [inline src/ansys/dpf/post/connectivity.py:31]: Parameters ?
```suggestion
        Parameters:
```

File: examples/01-Detailed-Examples/05-mesh-exploration.py (hunk 15 above)
Comment (PProfizi) [reply examples/01-Detailed-Examples/05-mesh-exploration.py:16]: @cbellot000 yes, I just added examples in dosctrings of post.Mesh
Will be updating the example to make a proper one

File: src/ansys/dpf/post/elements.py [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+        """Constructs an ElementType from an existing descriptor or enum id."""
+        self._el_desc = arg
+        if isinstance(arg, int):
+            self._el_desc = dpf.element_types.descriptor(arg)
+
+        if not isinstance(self._el_desc, dpf.ElementDescriptor):
+            raise TypeError("Given argument is not an int nor an ElementDescriptor")
+
+    @property
+    def enum_id(self) -> dpf.element_types:
```
Comment (cbellot000) [inline src/ansys/dpf/post/elements.py:27]: any other naming that enum_id ?

File: src/ansys/dpf/post/named_selection.py [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+        """Length of the list of IDs."""
+        return self._scoping.size
+
+    def deep_copy(self, server=None) -> NamedSelection:
+        """Create a deep copy of the underlying scoping's data on a given server."""
+        new_scoping = self._scoping.deep_copy(server)
+        new_name = copy.copy(self._name)
+        return NamedSelection(name=new_name, scoping=new_scoping)
+
+    def as_local_scoping(self) -> NamedSelection:
```
Comment (cbellot000) [inline src/ansys/dpf/post/named_selection.py:147]: is that needed?

File: src/ansys/dpf/post/named_selection.py [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+    def location(self) -> str:
+        """Location of the IDs as a string."""
+        return self._scoping.location
+
+    @property
+    def size(self) -> int:
+        """Length of the list of IDs."""
+        return self._scoping.size
+
+    def deep_copy(self, server=None) -> NamedSelection:
```
Comment (cbellot000) [inline src/ansys/dpf/post/named_selection.py:141]: this is not performant at the moment in core side as discussed with @rafacanton, do we need on the post side?

File: src/ansys/dpf/post/mesh.py [hunk 20]
```diff
@@ -21,21 +34,419 @@ def __str__(self):
...
+
+        Examples
+        --------
+        >>> from ansys.dpf import post
+        >>> from ansys.dpf.post import examples
+        >>> simulation = post.load_simulation(examples.static_rst)
+        >>> print(simulation.mesh.num_nodes) # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+        81
+        """
+        return len(self.node_ids)
```
Comment (cbellot000) [inline src/ansys/dpf/post/mesh.py:80]: can we use the core API for the number of nodes here please? We will eventually have a better implementation in core and I'ld like to inherit from it here

Comment (cbellot000) [reply src/ansys/dpf/post/mesh.py:80]: (same for elements)

File: src/ansys/dpf/post/elements.py (hunk 17 above)
Comment (cbellot000) [reply src/ansys/dpf/post/elements.py:27]: you are using "elem_type_id" in the DataFrame

File: src/ansys/dpf/post/simulation.py [hunk 21]
```diff
@@ -88,6 +90,18 @@ def release_streams(self):
         """Release the streams to data files if any is active."""
         self._model.metadata.release_streams()
 
+    @property
+    def mesh_info(self):
+        """Return available mesh information."""
+        mesh_selection_manager_provider_op = (
+            dpf.operators.metadata.mesh_selection_manager_provider(
```
Comment (cbellot000) [inline src/ansys/dpf/post/simulation.py:97]: why are we using the mesh_selection_manager_provider ?

File: src/ansys/dpf/post/elements.py [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+        """Returns the number of corner nodes."""
+        return self._el_desc.n_corner_nodes
+
+    @property
+    def num_mid_nodes(self) -> int:
+        """Returns the number of middle nodes."""
+        return self._el_desc.n_mid_nodes
+
+    @property
+    def num_nodes(self) -> int:
```
Comment (rafacanton) [inline src/ansys/dpf/post/elements.py:57]: How do these methods behave with Polys? (where we don't know the numbers)

Comment (PProfizi) [reply src/ansys/dpf/post/elements.py:57]: @rafacanton it's my next step, to adapt these to fluid meshes, in a next PR

**Review by rafacanton — APPROVED** [review]
